ipv8/dht/discovery.py

```diff
@@ -0,0 +1,185 @@
+from twisted.python.failure import Failure
+
+from .community import DHTCommunity, Request, PING_INTERVAL, TARGET_NODES, \
+                                            gatherResponses, MAX_NODES_IN_FIND
+from .routing import NODE_STATUS_BAD, Node
+from .payload import StorePeerRequestPayload, StorePeerResponsePayload, \
+                     ConnectPeerRequestPayload, ConnectPeerResponsePayload, \
+                     PingRequestPayload, PingResponsePayload
+
+MSG_STORE_PEER_REQUEST = 13
+MSG_STORE_PEER_RESPONSE = 14
+MSG_CONNECT_PEER_REQUEST = 15
+MSG_CONNECT_PEER_RESPONSE = 16
+
+
+class DHTDiscoveryCommunity(DHTCommunity):
+    """
+    Community for discovering peers that are behind NAT.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(DHTDiscoveryCommunity, self).__init__(*args, **kwargs)
+
+        self.store = defaultdict(list)
+        self.store_for_me = defaultdict(list)
+
+        self.decode_map.update({
+            chr(MSG_STORE_PEER_REQUEST): self.on_store_peer_request,
+            chr(MSG_STORE_PEER_RESPONSE): self.on_store_peer_response,
+            chr(MSG_CONNECT_PEER_REQUEST): self.on_connect_peer_request,
+            chr(MSG_CONNECT_PEER_RESPONSE): self.on_connect_peer_response,
+        })
+
+        self.register_task('store_peer', LoopingCall(self.store_peer)).start(30, now=False)
+
+    def on_ping_request(self, source_address, data):
+        super(DHTDiscoveryCommunity, self).on_ping_request(source_address, data)
+        auth, _, _ = self._ez_unpack_auth(PingRequestPayload, data)
+        for _, nodes in self.store.iteritems():
+            for node in nodes:
+                if node.public_key.key_to_bin() == auth.public_key_bin:
+                    node.last_query = time.time()
+
+    def on_ping_response(self, source_address, data):
+        super(DHTDiscoveryCommunity, self).on_ping_response(source_address, data)
+        auth, _, _ = self._ez_unpack_auth(PingResponsePayload, data)
+        for _, nodes in self.store_for_me.iteritems():
+            for node in nodes:
+                if node.public_key.key_to_bin() == auth.public_key_bin:
+                    node.last_response = time.time()
+
+    def store_peer(self):
+        # Do we already have enough peers storing our address?
+        if len(self.store_for_me) >= TARGET_NODES / 2:
+            return
+
+        key = self.my_peer.mid
+        return self.find_nodes(key).addCallback(lambda nodes: self.send_store_peer_request(key, nodes[:TARGET_NODES])) \
+                                   .addErrback(lambda _: None)
+
+    def send_store_peer_request(self, key, nodes):
+        # Create new objects to avoid problem with the nodes also being in the routing table
+        nodes = [Node(node.key, node.address) for node in nodes if node not in self.store_for_me[key]]
+
+        if not nodes:
+            return fail(Failure(RuntimeError('No nodes found for storing peer')))
+
+        deferreds = []
+        for node in nodes:
```

Comment:
I have a feeling I saw a block of code exactly like this in the parent community... Maybe you can make it a method, and reuse it? For DRY's sake :wink: